resolver/src/p2p/peer/actor.rs

```diff
@@ -0,0 +1,303 @@
+                            let Some(key) = self.fetcher.got_response(&peer, id, false) else {
+                                continue;
+                            };
+
+                            // The peer did not have the data, so we need to try again
+                            self.fetcher.fetch(&mut sender, key, true).await.unwrap(); // Unwrap must be safe
+                        },
+                    }
+                },
+
+                // Handle held deadline
+                _ = held_deadline => {
+                    let key = self.fetcher.pop_held();
+                    debug!(?key, "retrying");
+                    self.fetcher.fetch(&mut sender, key, true).await.unwrap(); // Unwrap must be safe
+                },
+
+                // Handle open deadline
+                _ = open_deadline => {
+                    if let Some(key) = self.fetcher.pop_open() {
+                        debug!(?key, "requester timeout");
+                        self.fetcher.fetch(&mut sender, key, true).await.unwrap(); // Unwrap must be safe
+                    }
+                },
+            }
+        }
+    }
+
+    /// Handles the case where the application responds to a request from an external peer.
+    async fn handle_serve(
+        sender: &mut NetS,
+        peer: C::PublicKey,
+        id: u64,
+        response: Result<Value, oneshot::Canceled>,
+    ) {
+        // Encode message
+        let msg = wire::PeerMsg {
+            id,
+            payload: response.ok().map(Payload::Response),
+        }
+        .encode_to_vec()
+        .into();
+
+        // Send message to peer
+        let result = sender.send(Recipients::One(peer.clone()), msg, false).await;
+
+        // Log result, but do not handle errors
+        match result {
+            Err(err) => error!(?err, ?peer, ?id, "serve send failed"),
+            Ok(to) if to.is_empty() => warn!(?peer, ?id, "serve send failed"),
+            Ok(_) => debug!(?peer, ?id, "serve sent"),
+        };
+    }
+
+    /// Handles the case where a peer sends a request to this peer.
+    fn handle_request(&mut self, peer: C::PublicKey, id: u64, request: Key) {
+        // If the peer is not allowed to request, drop the request
+        if !self.director.is_peer(&peer) {
+            warn!(?peer, ?id, "dropping request: peer not allowed");
+            return;
+        }
+
+        // If there are too many requests from this peer, drop the request
+        if self.rate_limiter.check_key(&peer).is_err() {
+            warn!(?peer, ?id, "dropping request: rate limit exceeded");
+            return;
+        }
+
+        // If there are too many pending requests, drop the request
+        // TODO: consider sending a failure response (?)
+        let n = self.serves.len();
+        if n >= self.serve_concurrent {
+            warn!(?peer, ?id, ?n, "dropping request: too many pending");
+            return;
+        }
```

Comment:
Do we want to consider sending an error response instead of just dropping the request?
We should be insulated by the `rate_limiter` at this point so it's probably okay (?)

Reply:
See my other message about redundant rate limiters.

Reply:
As for serve_concurrent being exhausted, I think we return an empty response.

Normally, you shouldn't do anything when "overloaded" but it seems ok because we are already rate limiting the P2P before this point.

Reply:
On second thought, I think we should consider removing this too and relying entirely on the channel rate limiting.

If we are using this, it just means that RL is misconfigured.

Reply:
To clarify, we still need to pass the channel RL to the requester to avoid accidentally rate limiting ourself (and it looks like you do that 👍 ).

Reply:
Removed `serve_concurrent` and `fetch_max_outstanding`